plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogAccess.java

```diff
@@ -295,12 +296,16 @@ private Stream<AddFileEntry> activeAddEntries(Stream<DeltaLakeTransactionLogEntr
             AddFileEntry addEntry = deltaLakeTransactionLogEntry.getAdd();
             if (addEntry != null) {
                 activeJsonEntries.put(addEntry.getPath(), addEntry);
+                addEntry.getDeletionVector().ifPresent(deletionVector -> dependOnDeletionVector.add(addEntry.getPath()));
             }
 
             RemoveFileEntry removeEntry = deltaLakeTransactionLogEntry.getRemove();
             if (removeEntry != null) {
-                activeJsonEntries.remove(removeEntry.getPath());
                 removedFiles.add(removeEntry.getPath());
+                if (!dependOnDeletionVector.contains(removeEntry.getPath())) {
+                    // Deletion vector registers both 'add' & 'remove' entries and 'add' entry should be kept
```

Comment:
let me recap what i understand

- when we do row-level delete from a file P
  -  a `remove` entry is emitted, with the file path P
  - an `add` entry is emitted, with the file path P and deletion vector D
- within transaction (JSON) object, these `remove` and `add` entries can be in any order

am i understanding this correct?

if so, then

- the logic handling remove-after-add should be scoped within transaction. currently `dependOnDeletionVector` spans multiple transaction entries, so if subsequent transaction has `remove P` , it is going to be ignored

- there are probably easier ways to handle this problem; for example if we operated on _transactions_ instead of individual transaction entries, we would simply process removals before additions
  - i think all we need is to change `io.trino.plugin.deltalake.transactionlog.checkpoint.TransactionLogTail#entries` to be grouped by transaction

Reply:
+1 for grouping transaction log entries by transaction - could be done in a prep PR